extensions/amp-ad-network-doubleclick-impl/0.1/sra-utils.js

```diff
@@ -0,0 +1,292 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from '../../../src/log';
+import {getEnclosingContainerTypes} from '../../../ads/google/a4a/utils';
+import {
+  isInManualExperiment,
+} from '../../../ads/google/a4a/traffic-experiments';
+
+/**
+ * @const {string}
+ * @visibleForTesting
+ */
+export const TFCD = 'tagForChildDirectedTreatment';
+
+/** @private {!Array<function(!Array<!./amp-ad-network-doubleclick-impl.AmpAdNetworkDoubleclickImpl>):?Object<string,string>>} */
+const SRA_JOINERS = [
+  combineInventoryUnits, getCookieOptOut, getAdks, getSizes, getTfcd, isAdTest,
+  getTargetingAndExclusions, getExperimentIds, getIdentity, getForceSafeframe,
+  getPageOffsets, getContainers];
+
+/**
+  * @param {!Array<!./amp-ad-network-doubleclick-impl.AmpAdNetworkDoubleclickImpl>} impls
+  * @return {!Object<string, *>}
+  */
+export function constructSRABlockParameters(impls) {
+  const parameters = {'output': 'ldjh', 'impl': 'fifs'};
+  SRA_JOINERS.forEach(joiner => Object.assign(parameters, joiner(impls)));
+  return parameters;
+}
+
+/**
+  * @param {!Array<!./amp-ad-network-doubleclick-impl.AmpAdNetworkDoubleclickImpl>} impls
+  * @param {function(!./amp-ad-network-doubleclick-impl.AmpAdNetworkDoubleclickImpl):?T} extractFn
+  * @return {?T} value of first instance with non-null/undefined value or null
+  *    if none can be found
+  * @template T
+  * @private
+  */
+function getFirstInstanceValue_(impls, extractFn) {
+  for (let i = 0; i < impls.length; i++) {
+    const val = extractFn(impls[i]);
+    if (val) {
+      return val;
+    }
+  }
+  return null;
+}
+
+/**
+ * Combines inventory unit paths for multiple blocks by building list of
+ * unique path parts in iu_parts and then comma separated list of block
+ * paths using index into iu_parts list.
+ * @param {!Array<!./amp-ad-network-doubleclick-impl.AmpAdNetworkDoubleclickImpl>} impls
+ * @return {?Object<string,string>}
+ * @visibleForTesting
+ */
+export function combineInventoryUnits(impls) {
+  const uniqueIuNames = {};
+  let uniqueIuNamesCount = 0;
+  const prevIusEncoded = [];
+  impls.forEach(instance => {
+    const iu = dev().assert(instance.element.getAttribute('data-slot'));
+    const componentNames = (iu || '').split('/');
+    const encodedNames = [];
+    for (let i = 0; i < componentNames.length; i++) {
+      if (componentNames[i] == '') {
```

Comment:
This code seems really overcomplicated to me, when all you want in the end is an array of unique IU names in the order you saw them. I think you can probably just use Object.keys, or to be extra safe, you can have an array that you build as you go, and you use the obj as a lookup table before you put in a new entry. Then you can also get rid of the sorting at the end, which you really don't need unless I'm really missing something here. Could look like this: 

```
export function combineInventoryUnits(impls) {
  const uniqueIuNames = {};
  const uniqueIuArray = [];
  const prevIusEncoded = [];
  impls.forEach(instance => {
    const iu = dev().assert(instance.element.getAttribute('data-slot'));
    const componentNames = (iu || '').split('/');
    const encodedNames = [];
    for (let i = 0; i < componentNames.length; i++) {
      if (componentNames[i] == '') {
        continue;
      }
      if (!uniqueIuNames[componentNames[i]]) {
        uniqueIuNames[componentNames[i]] = "unused";
        uniqueIuArray.push(componentNames[i]);
      }
      encodedNames.push(uniqueIuArray.length);
    }
    prevIusEncoded.push(encodedNames.join('/'));
  });
  return {
    'iu_parts': uniqueIuArray.join(),
    'enc_prev_ius': prevIusEncoded.join(),
  };
}

Reply:
We can't actually rely on object.keys being in the same order as insertion (I found this out the hard way!  apparently keys are sometimes sorted if numeric... its odd).  In order to make previous component lookup O(1), I store component to position map.  I could remove the sort at the end and replace it with an array, just not sure it really matters.